learning_phase/chat_sample/client.go

```diff
@@ -0,0 +1,88 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"os"
+	"strings"
+)
+
+//various commands numbered from 0, 1, 2, 3...
+const (
+	cmdauth = iota
+	cmdpm
+	cmdmsg
+	cmdquit
+)
+
+type command struct {
+	id     int
+	client *client
+	args   []string
+}
+
+type client struct {
+	conn     net.Conn
+	address  string
+	username string
+	commands chan<- command //incoming channel
+}
+
+func (c *client) readInput(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done(): //when cancel is called
+			log.Printf("Terminating Clients")
+			os.Exit(0)
+		default:
+			msg, err := bufio.NewReader(c.conn).ReadString('\n')
+			if err != nil {
+				return
+			}
+			//parse the string
+			msg = strings.Trim(msg, "\r\n")
+
+			args := strings.Split(msg, " ")
```

Comment:
nitpick. we did discuss using whitespace as a delimiter, but it'd be a problem if say we wanted to allow spaces in people's usernames. Maybe a better delimiter would be a less-used special character like ~ (still not perfect, but can be improved by adding a special encoding for that character in case the user enters it in their message)